sdl_android/src/main/java/com/smartdevicelink/api/FileManager.java

```diff
@@ -0,0 +1,428 @@
+import android.content.res.Resources;
+import android.net.Uri;
+import android.util.Log;
+import android.util.SparseArray;
+
+import com.smartdevicelink.api.datatypes.SdlArtwork;
+import com.smartdevicelink.api.datatypes.SdlFile;
+import com.smartdevicelink.proxy.RPCRequest;
+import com.smartdevicelink.proxy.RPCResponse;
+import com.smartdevicelink.proxy.interfaces.ISdl;
+import com.smartdevicelink.proxy.rpc.DeleteFile;
+import com.smartdevicelink.proxy.rpc.ListFiles;
+import com.smartdevicelink.proxy.rpc.ListFilesResponse;
+import com.smartdevicelink.proxy.rpc.PutFile;
+import com.smartdevicelink.proxy.rpc.enums.Result;
+import com.smartdevicelink.proxy.rpc.listeners.OnMultipleRequestListener;
+import com.smartdevicelink.proxy.rpc.listeners.OnRPCResponseListener;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * <strong>FileManager</strong> <br>
+ *
+ * Note: This class must be accessed through the SdlManager. Do not instantiate it by itself. <br>
+ *
+ * The SDLFileManager uploads files and keeps track of all the uploaded files names during a session. <br>
+ *
+ * We need to add the following struct: SDLFile<br>
+ *
+ * It is broken down to these areas: <br>
+ *
+ * 1. Getters <br>
+ * 2. Deletion methods <br>
+ * 3. Uploading Files / Artwork
+ */
+public class FileManager extends BaseSubManager {
+
+	private static String TAG = "FileManager";
+	private List<String> remoteFiles, uploadedEphemeralFileNames;
+	private WeakReference<Context> context;
+
+	FileManager(ISdl internalInterface, Context context) {
+
+		// setup
+		super(internalInterface);
+		this.context = new WeakReference<>(context);
+
+		// prepare manager - don't set state to ready until we have list of files
+		retrieveRemoteFiles();
+		uploadedEphemeralFileNames = new ArrayList<>();
+	}
+
+	// GETTERS
+
+	/**
+	 * Returns a list of file names currently residing on core
+	 * @return List<String> of remote file names
+	 */
+	public List<String> getRemoteFileNames() {
+		if (getState() != BaseSubManager.READY){
+			// error and dont return list
+			throw new IllegalArgumentException("FileManager is not READY");
+		}
+		// return list (this is synchronous at this point)
+		return remoteFiles;
+	}
+
+	private void retrieveRemoteFiles(){
+		remoteFiles = new ArrayList<>();
+		// hold list in remoteFiles class var
+		ListFiles listFiles = new ListFiles();
+		listFiles.setOnRPCResponseListener(new OnRPCResponseListener() {
+			@Override
+			public void onResponse(int correlationId, RPCResponse response) {
+				if(response.getSuccess()){
+					if(((ListFilesResponse) response).getFilenames() != null){
+						remoteFiles.addAll(((ListFilesResponse) response).getFilenames());
+					}
+					// on callback set manager to ready state
+					transitionToState(BaseSubManager.READY);
+				}else{
```

Comment:
This case will actually never happen. `onResponse` is only called when the response was successful. There is another method to override `onError` that is called when issues occur.